peekingduck/pipeline/nodes/model/jde_mot/jde_files/utils/utils.py

```diff
@@ -0,0 +1,256 @@
+
+    Returns:
+        List[torch.Tensor | None]: Returns detections with shape;
+            (x1, y1, x2, y2, object_conf, class_score, class_pred)
+    """
+    output = [None for _ in range(len(prediction))]
+    for image_i, pred in enumerate(prediction):
+        # Filter out confidence scores below threshold
+        # Get score and class with highest confidence
+
+        v = pred[:, 4] > conf_thres
+        v = v.nonzero().squeeze()
+        if len(v.shape) == 0:
+            v = v.unsqueeze(0)
+
+        pred = pred[v]
+
+        # If none are remaining => process next image
+        nP = pred.shape[0]
+        if not nP:
+            continue
+        # From (center x, center y, width, height) to (x1, y1, x2, y2)
+        pred[:, :4] = xywh2xyxy(pred[:, :4])
+
+        # Non-maximum suppression
+        nms_indices = nms(pred[:, :4], pred[:, 4], nms_thres)
+        det_max = pred[nms_indices]
+
+        if len(det_max) > 0:
+            # Add max detections to outputs
+            output[image_i] = (
+                det_max
+                if output[image_i] is None
+                else torch.cat((output[image_i], det_max))  # type: ignore
+            )
+    return output
+
+
+def xywh2xyxy(x: torch.Tensor) -> torch.Tensor:
+    """Convert bounding box format from [x, y, w, h] to [x1, y1, x2, y2].
+    (x, y) are coordinates of center. (x1, y1) and (x2, y2) are coordinates
+    of bottom left and top right respectively.
+
+    Args:
+        x (torch.Tensor): Tensor with bounding boxes of format
+            `(center x, center y, width, height)`.
+
+    Returns:
+        torch.Tensor: Tensor with bounding boxes of format
+            `(top left x, top left y, bottom right x, bottom right y)`.
+    """
+    y = torch.zeros_like(x) if x.dtype is torch.float32 else np.zeros_like(x)
+    y[:, 0] = x[:, 0] - x[:, 2] / 2  # Bottom left x
+    y[:, 1] = x[:, 1] - x[:, 3] / 2  # Bottom left y
+    y[:, 2] = x[:, 0] + x[:, 2] / 2  # Top right x
+    y[:, 3] = x[:, 1] + x[:, 3] / 2  # Top right y
+    return y
+
+
+def generate_anchor(
+    nGh: int, nGw: int, anchor_wh: torch.Tensor, device: torch.device
+) -> torch.Tensor:
+    """Generate anchor bounding box for prediction.
+
+    Args:
+        nGh (int): Number of grid height.
+        nGw (int): Number of grid width.
+        anchor_wh (torch.Tensor): Anchor for bounding boc width, height.
+        device (torch.device): Device type being used. "cuda" or "cpu".
+
+    Returns:
+        torch.Tensor: Prediction bounding box anchors.
+    """
+    nA = len(anchor_wh)
+    yy, xx = torch.meshgrid(torch.arange(nGh), torch.arange(nGw))
+    xx = xx.to(device)
+    yy = yy.to(device)
+
+    mesh = torch.stack([xx, yy], dim=0)  # Shape 2, nGh, nGw
+    mesh = mesh.unsqueeze(0).repeat(nA, 1, 1, 1).float()  # Shape nA x 2 x nGh x nGw
+    anchor_offset_mesh = (
+        anchor_wh.unsqueeze(-1).unsqueeze(-1).repeat(1, 1, nGh, nGw)
+    )  # Shape nA x 2 x nGh x nGw
```

Comment:
avoid inline comment if it causes code to wrap

Reply:
Updated